cg/apps/balsamic/fastq.py

```diff
@@ -0,0 +1,151 @@
+
+        size_before = FastqFileConcatenator().size_before(files)
+        size_after = FastqFileConcatenator().size_after(concat_file)
+
+        try:
+            FastqFileConcatenator().assert_file_sizes(size_before, size_after)
+        except AssertionError as error:
+            logger.warning(error)
+
+    @staticmethod
+    def size_before(files: List):
+        """returns the total size of the linked fastq files before concatenation"""
+
+        return sum([os.stat(f).st_size for f in files])
+
+    @staticmethod
+    def size_after(concat_file):
+        """returns the size of the concatenated fastq files"""
+
+        return os.stat(concat_file).st_size
+
+    @staticmethod
+    def assert_file_sizes(size_before, size_after):
+        """asserts the file sizes before and after concatenation. If the file sizes differ by more
+        than 1 percent, throw an exception"""
+        msg = (
+            f"Warning: Large file size difference after concatenation!"
+            f"Before: {size_before} -> after: {size_after}"
+        )
+
+        assert abs(size_before - size_after) / size_before <= 0.01, msg
+        logger.info('Concatenation file size check successful!')
+
+    @staticmethod
+    def display_files(files: List, concat_file):
+        """display file names for logging purposes"""
+
+        concat_file_name = Path(concat_file).name
+        msg = f"Concatenating: {', '.join(Path(file_).name for file_ in files)} -> " \
+            f"{concat_file_name}"
+
+        return msg
+
+
+class FastqHandler:
+    """Handles fastq file linking"""
+
+    def __init__(self, config):
+        self.root_dir = config['balsamic']['root']
+
+    def link(self, family: str, sample: str, files: List):
+        """Link FASTQ files for a balsamic sample.
+        Shall be linked to /mnt/hds/proj/bionfo/BALSAMIC_ANALYSIS/case-id/fastq/"""
+
+        wrk_dir = Path(f'{self.root_dir}/{family}/fastq')
+
+        wrk_dir.mkdir(parents=True, exist_ok=True)
+
+        destination_paths_r1 = list()
+        destination_paths_r2 = list()
+
+        sorted_files = sorted(files, key=lambda k: k['path'])
+
+        for fastq_data in sorted_files:
+            fastq_path = Path(fastq_data['path'])
+            fastq_name = FastqFileNameCreator.create(
+                lane=fastq_data['lane'],
+                flowcell=fastq_data['flowcell'],
+                sample=sample,
+                read=fastq_data['read'],
+                undetermined=fastq_data['undetermined'],
+            )
+
+            destination_path = wrk_dir / fastq_name
+
+            if fastq_data['read'] == 1:
+                destination_paths_r1.append(destination_path)
+                concatenated_filename_r1 = f"concatenated_{'_'.join(fastq_name.split('_')[-4:])}"
+            else:
+                destination_paths_r2.append(destination_path)
+                concatenated_filename_r2 = f"concatenated_{'_'.join(fastq_name.split('_')[-4:])}"
```

Comment:
@patrikgrenfeldt  refactor?

Reply:
@barrystokman could you have a look at the refactoring I did?